### ISSUE TITLE
IndexOutOfBoundsException

### ISSUE THREAD

**Issue by JimVanHorn** [popiel/firstNofM#1]
I've been working with your presentation code.  Looking at the firstNCompletedOfM() code I wondered what might happen if two threads were to get past the "if" check before accessing the AtomicInteger.  To test this out I made the following mod:

  val last = promises.last
  val completed = new AtomicInteger(0)
  val handler: Try[T] => Unit = { result =>
    if (!last.isCompleted) {
      Thread.sleep(Random.nextInt(2000)) /_test_/
      promises(completed.getAndIncrement) success result
    }
  }

Running this got the IndexOutOfBoundsException.  Is this an unrealistic simulation?

Here's the mod I made to fix it:

//  val last = promises.last
  val counter = new AtomicInteger(0)
  val handler: Try[T] => Unit = { result =>
    val indx = counter.getAndIncrement
    if (indx < promises.size) {
      Thread.sleep(Random.nextInt(2000)) /_test_/
      promises(indx) success result
    }
  }

This appears to work.  I then looked at the real code, and was able to recreate the issue, but a very different fix suggested itself:

  val promises = Array.fill(futures.size)(Promise[Try[T]]())
  val index = new java.util.concurrent.atomic.AtomicInteger(0)
  val handler: Try[T] => Unit = { promises(index.getAndIncrement) success _ }
  futures foreach { _ onComplete handler }
  val builder = cbf1(futures)
  val actual = promises.size min count
  builder.sizeHint(actual)
  (promises take actual) foreach { builder += _.future }
  Future.sequence(builder.result)

But I must admit I haven't studied the CanBuildFrom utility so maybe this mod misses an important code feature.

-jvh

**Comment by popiel** [comment]
Good catch.  I think this may be a pervasive issue across the filtered variants, too, so I'm going to think a bit on a good way to fix before posting.

**Comment by popiel** [comment]
Fixed by adding in array bounds checks.  It was the cleanest of all the various approaches...

**Comment by JimVanHorn** [comment]
Hi Alex,

Looks good.

Just FYI: A few days ago I ran across a site 
(pavelfatin.com/scala-collections-tips-and-tricks) that points out that 
we seldom need to explicitly check for index bounds.  The trick is to 
use lift (news to me) to extract an Option which can be unwrapped using 
flatMap, foreach, etc.

I plugged it in to your code and it seems to work well, passing all my 
(limited) tests.

   val handler: Try[T] => Unit = { t =>
     promises.lift(index.getAndIncrement).foreach(_ success t)
   }

Like I say, just an FYI,
-jvh
/Jim Van Horn
Seattle, WA USA/
